From 2abdb1e89d7a3055acc5e5a5fd98bc18b8a07468 Mon Sep 17 00:00:00 2001
From: Hanna Hawa <hannah@marvell.com>
Date: Tue, 10 Jun 2014 18:41:54 +0300
Subject: [PATCH 0035/1240] a8k: pxp: add workaround for design bug in Aurora2

	- added read before/after every register write
	- bug reason: the CPU send two transactions for the memory with the
	  same id. The CPU send the second transaction without receiving the first one

Change-Id: I2b1411531ba364190e969dc17efd6450e681d8ed
Signed-off-by: Hanna Hawa <hannah@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/8458
Tested-by: Star_Automation <star@marvell.com>
Reviewed-by: Yehuda Yitschak <yehuday@marvell.com>
---
 arch/arm/include/asm/io.h | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/arch/arm/include/asm/io.h b/arch/arm/include/asm/io.h
index dc623e8..0fb258f 100644
--- a/arch/arm/include/asm/io.h
+++ b/arch/arm/include/asm/io.h
@@ -72,10 +72,16 @@ static inline phys_addr_t virt_to_phys(void * vaddr)
 #define __arch_getl(a)			(*(volatile unsigned int *)(a))
 #define __arch_getq(a)			(*(volatile unsigned long long *)(a))
 
+#if CONFIG_PALLADIUM
+#define __arch_putb(v,a)		{__arch_getb(a); (*(volatile unsigned char *)(a) = (v)); __arch_getb(a);}
+#define __arch_putw(v,a)		{__arch_getw(a); (*(volatile unsigned short *)(a) = (v)); __arch_getw(a);}
+#define __arch_putl(v,a)		{__arch_getl(a); (*(volatile unsigned int *)(a) = (v)); __arch_getl(a);}
+#else
 #define __arch_putb(v,a)		(*(volatile unsigned char *)(a) = (v))
 #define __arch_putw(v,a)		(*(volatile unsigned short *)(a) = (v))
 #define __arch_putl(v,a)		(*(volatile unsigned int *)(a) = (v))
 #define __arch_putq(v,a)		(*(volatile unsigned long long *)(a) = (v))
+#endif
 
 static inline void __raw_writesb(unsigned long addr, const void *data,
 				 int bytelen)
-- 
1.9.1

